File: app/src/main/java/dev/ragnarok/fenrir/dialog/SelectChairsDialog.java
package dev.ragnarok.fenrir.dialog;

import android.app.Dialog;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import java.util.ArrayList;
import java.util.List;

import dev.ragnarok.fenrir.Extra;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.adapter.vkdatabase.ChairsAdapter;
import dev.ragnarok.fenrir.dialog.base.AccountDependencyDialogFragment;
import dev.ragnarok.fenrir.domain.IDatabaseInteractor;
import dev.ragnarok.fenrir.domain.InteractorFactory;
import dev.ragnarok.fenrir.model.database.Chair;
import dev.ragnarok.fenrir.util.RxUtils;

public class SelectChairsDialog extends AccountDependencyDialogFragment implements ChairsAdapter.Listener {

    public static final String REQUEST_CODE_CHAIRS = "request_chairs";
    private static final int COUNT_PER_REQUEST = 1000;
    private int mAccountId;
    private int facultyId;
    private ArrayList<Chair> mData;
    private ChairsAdapter mAdapter;
    private IDatabaseInteractor mDatabaseInteractor;

    public static SelectChairsDialog newInstance(int aid, int facultyId, Bundle additional) {
        Bundle args = additional == null ? new Bundle() : additional;
        args.putInt(Extra.FACULTY_ID, facultyId);
        args.putInt(Extra.ACCOUNT_ID, aid);
        SelectChairsDialog selectCityDialog = new SelectChairsDialog();
        selectCityDialog.setArguments(args);
        return selectCityDialog;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mAccountId = requireArguments().getInt(Extra.ACCOUNT_ID);
        mDatabaseInteractor = InteractorFactory.createDatabaseInteractor();
        facultyId = requireArguments().getInt(Extra.FACULTY_ID);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        View root = View.inflate(requireActivity(), R.layout.dialog_simple_recycler_view, null);
        RecyclerView mRecyclerView = root.findViewById(R.id.list);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(requireActivity(), RecyclerView.VERTICAL, false));

        boolean firstRun = false;
        if (mData == null) {
            mData = new ArrayList<>();
            firstRun = true;
        }

        mAdapter = new ChairsAdapter(requireActivity(), mData);
        mAdapter.setListener(this);
        mRecyclerView.setAdapter(mAdapter);

        if (firstRun) {
            request(0);
        }

        return new MaterialAlertDialogBuilder(requireActivity())
                .setTitle(R.string.chair)
                .setView(root)
                .setNegativeButton(R.string.button_cancel, null)
                .create();
    }

    private void request(int offset) {
        appendDisposable(mDatabaseInteractor.getChairs(mAccountId, facultyId, COUNT_PER_REQUEST, offset)
                .compose(RxUtils.applySingleIOToMainSchedulers())
                .subscribe(chairs -> onDataReceived(offset, chairs), throwable -> {
                }));
    }

    private void onDataReceived(int offset, List<Chair> chairs) {
        if (offset == 0) {
            mData.clear();
        }

        mData.addAll(chairs);
        mAdapter.notifyDataSetChanged();
    }

    @Override
    public void onClick(Chair chair) {
        Bundle intent = new Bundle();
        intent.putParcelable(Extra.CHAIR, chair);
        intent.putInt(Extra.ID, chair.getId());
        intent.putString(Extra.TITLE, chair.getTitle());

        if (getArguments() != null) {
            intent.putAll(getArguments());
        }
        getParentFragmentManager().setFragmentResult(REQUEST_CODE_CHAIRS, intent);
        dismiss();
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/dialog/selectchairs/SelectChairsDialog.kt
package dev.ragnarok.fenrir.dialog.selectchairs

import android.app.Dialog
import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import dev.ragnarok.fenrir.Extra
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.dialog.base.AccountDependencyDialogFragment
import dev.ragnarok.fenrir.domain.IDatabaseInteractor
import dev.ragnarok.fenrir.domain.InteractorFactory
import dev.ragnarok.fenrir.fragment.search.filteredit.FilterEditFragment
import dev.ragnarok.fenrir.fromIOToMain
import dev.ragnarok.fenrir.model.database.Chair
import java.util.Collections

class SelectChairsDialog : AccountDependencyDialogFragment(), ChairsAdapter.Listener {
    private var chairId = 0
    private var mData: ArrayList<Chair>? = null
    private var mAdapter: ChairsAdapter? = null
    private val mDatabaseInteractor: IDatabaseInteractor =
        InteractorFactory.createDatabaseInteractor()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        chairId = requireArguments().getInt(Extra.CHAIR_ID)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val root = View.inflate(requireActivity(), R.layout.dialog_simple_recycler_view, null)
        val mRecyclerView: RecyclerView = root.findViewById(R.id.list)
        mRecyclerView.layoutManager =
            LinearLayoutManager(requireActivity(), RecyclerView.VERTICAL, false)
        var firstRun = false
        if (mData == null) {
            mData = ArrayList()
            firstRun = true
        }
        mAdapter = ChairsAdapter(requireActivity(), mData ?: Collections.emptyList())
        mAdapter?.setListener(this)
        mRecyclerView.adapter = mAdapter
        if (firstRun) {
            request(0)
        }
        return MaterialAlertDialogBuilder(requireActivity())
            .setTitle(R.string.chair)
            .setView(root)
            .setNegativeButton(R.string.button_cancel, null)
            .create()
    }

    private fun request(offset: Int) {
        appendDisposable(
            mDatabaseInteractor.getChairs(accountId, chairId, COUNT_PER_REQUEST, offset)
                .fromIOToMain()
                .subscribe({ chairs ->
                    onDataReceived(
                        offset,
                        chairs
                    )
                }) { })
    }

    private fun onDataReceived(offset: Int, chairs: List<Chair>) {
        if (offset == 0) {
            mData?.clear()
        }
        mData?.addAll(chairs)
        mAdapter?.notifyDataSetChanged()
    }

    override fun onClick(chair: Chair) {
        val intent = Bundle()
        intent.putParcelable(Extra.CHAIR, chair)
        intent.putInt(Extra.ID, chair.id)
        intent.putString(Extra.TITLE, chair.title)
        if (arguments != null) {
            intent.putAll(arguments)
        }
        parentFragmentManager.setFragmentResult(FilterEditFragment.REQUEST_FILTER_OPTION, intent)
        dismiss()
    }

    companion object {
        private const val COUNT_PER_REQUEST = 1000
        fun newInstance(aid: Long, chairId: Int, additional: Bundle?): SelectChairsDialog {
            val args = additional ?: Bundle()
            args.putInt(Extra.CHAIR_ID, chairId)
            args.putLong(Extra.ACCOUNT_ID, aid)
            val selectCityDialog = SelectChairsDialog()
            selectCityDialog.arguments = args
            return selectCityDialog
        }
    }
}